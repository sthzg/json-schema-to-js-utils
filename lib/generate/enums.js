const { createWriteStream } = require('fs');
const {
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  FilterEnumsTransform,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  StdoutWritable,
  PrependToFileTransform,
  PrettierTransform,
} = require('../streams');
const { enumModule } = require('../templates');

module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  jsonSchemaReadable(schema)
    .pipe(new FilterPropertiesTransform())
    .pipe(new FilterEnumsTransform())
    .pipe(new BuildObjectPathsTransform())
    .pipe(new CompileToTemplateTransform(enumModule))
    .pipe(new MemorySinkTransform())
    .pipe(new PrependToFileTransform(DEFAULT_FILE_DOCSTRING))
    .pipe(new PrettierTransform(DEFAULT_PRETTIER_OPTIONS))
    .pipe(write);
};
