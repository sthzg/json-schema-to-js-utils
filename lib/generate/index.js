import generateEnumsModule from './enums';
import generateObjectPathsModule from './paths';
import generateSelectorsModule from './selectors';

export {
  generateEnumsModule,
  generateObjectPathsModule,
  generateSelectorsModule,
};
