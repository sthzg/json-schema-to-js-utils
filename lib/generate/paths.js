const { createWriteStream } = require('fs');
const { pipeline: pipelineFn } = require('stream');
const { promisify } = require('util');
const {
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  StdoutWritable,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  PrependToFileTransform,
  PrettierTransform,
} = require('../streams');
const { objectPathConstant } = require('../templates');

const pipeline = promisify(pipelineFn);

module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  return pipeline(
    jsonSchemaReadable(schema),
    new FilterPropertiesTransform(),
    new BuildObjectPathsTransform(),
    new CompileToTemplateTransform(objectPathConstant),
    new MemorySinkTransform(),
    new PrependToFileTransform(DEFAULT_FILE_DOCSTRING),
    new PrettierTransform(DEFAULT_PRETTIER_OPTIONS),
    write
  );
};
