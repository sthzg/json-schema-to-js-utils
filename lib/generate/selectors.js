const { createWriteStream } = require('fs');
const { pipeline: pipelineFn } = require('stream');
const { promisify } = require('util');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  StdoutWritable,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  PrettierTransform,
  PrependToFileTransform,
} = require('../streams');
const {
  DEFAULT_GET_IMPORT,
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const { selectorFunctionModule } = require('../templates');

const pipeline = promisify(pipelineFn);

/**
 * Generates selector functions for all values inside the JSON schema.
 *
 * Selectors are functions that pick a value from an object in a failsafe way or return undefined if
 * the property is not defined on the input object, e.g.
 *
 *     function byPersonAddressStreet(person) {
 *       return get(person, 'address.street');
 *     }
 */
module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  return pipeline(
    jsonSchemaReadable(schema),
    new FilterPropertiesTransform(),
    new BuildObjectPathsTransform(),
    new CompileToTemplateTransform(selectorFunctionModule),
    new MemorySinkTransform(),
    new PrependToFileTransform(DEFAULT_FILE_DOCSTRING, DEFAULT_GET_IMPORT),
    new PrettierTransform(DEFAULT_PRETTIER_OPTIONS),
    write
  );
};
