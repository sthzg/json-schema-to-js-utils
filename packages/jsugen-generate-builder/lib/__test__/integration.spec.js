import schema from '@sthzg/jsugen-core/lib/testResources/example.schema';
import generate from '../index';

test('generate-builder integration test', () => {
  expect.assertions(1);

  generate({ schema }).subscribe(output => {
    expect(output).toMatchSnapshot();
  });
});
