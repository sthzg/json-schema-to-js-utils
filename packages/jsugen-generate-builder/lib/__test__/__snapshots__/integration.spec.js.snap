// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-builder integration test 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import cloneDeep from 'lodash-es/cloneDeep';
import isUndefined from 'lodash-es/isUndefined';
import merge from 'lodash-es/merge';
import set from 'lodash-es/set';

const mergeSet = (targetObject, keyValueObject) =>
  (0, _merge.default)(
    (0, _cloneDeep.default)(targetObject),
    ...Object.entries(keyValueObject)
      .filter(([, value]) => !(0, _isUndefined.default)(value))
      .map(([path, value]) => (0, _set.default)({}, path, value)),
  );

const builder = (
  {
    ROOT_LEVEL_STRING,
    ROOT_LEVEL_BOOL,
    ROOT_LEVEL_NUMBER,
    ROOT_LEVEL_ENUM_ARRAY,
    ROOT_LEVEL_ARRAY_ARRAY,
    ROOT_LEVEL_FROM_DEFINITION_MEMBER_DEFINED_IN_DEFINITION,
    ROOT_LEVEL_FROM_DEFINITION_OBJECT,
    LEVEL_ONE_ANY_STRING,
    LEVEL_ONE_ANY_ENUM_ARRAY,
    LEVEL_ONE_ANY_ARRAY_ARRAY,
    LEVEL_ONE_ANY_BOOL,
    LEVEL_ONE_ANY_NUMBER,
    LEVEL_ONE_LEVEL_TWO_FOO,
    LEVEL_ONE_LEVEL_TWO_OBJECT,
    LEVEL_ONE_OBJECT,
  },
  initialData = {},
) =>
  mergeSet(initialData, {
    rootLevelString: ROOT_LEVEL_STRING,
    rootLevelBool: ROOT_LEVEL_BOOL,
    rootLevelNumber: ROOT_LEVEL_NUMBER,
    rootLevelEnumArray: ROOT_LEVEL_ENUM_ARRAY,
    rootLevelArray: ROOT_LEVEL_ARRAY_ARRAY,
    'rootLevelFromDefinition.memberDefinedInDefinition': ROOT_LEVEL_FROM_DEFINITION_MEMBER_DEFINED_IN_DEFINITION,
    rootLevelFromDefinition: ROOT_LEVEL_FROM_DEFINITION_OBJECT,
    'levelOne.anyString': LEVEL_ONE_ANY_STRING,
    'levelOne.anyEnumArray': LEVEL_ONE_ANY_ENUM_ARRAY,
    'levelOne.anyArray': LEVEL_ONE_ANY_ARRAY_ARRAY,
    'levelOne.anyBool': LEVEL_ONE_ANY_BOOL,
    'levelOne.anyNumber': LEVEL_ONE_ANY_NUMBER,
    'levelOne.levelTwo.foo': LEVEL_ONE_LEVEL_TWO_FOO,
    'levelOne.levelTwo': LEVEL_ONE_LEVEL_TWO_OBJECT,
    levelOne: LEVEL_ONE_OBJECT,
  });

export default builder;
"
`;
